File: Python/0091-decode-ways.py
# Time:  O(n)
# Space: O(1)

class Solution:
    def numDecodings(self, s: str) -> int:
        if len(s) == 0 or s[0] == '0':
            return 0
        prev, prev_prev = 1, 1

        for i in range(1, len(s)):
            cur = 0
            if s[i] != '0':
                cur = prev
            if s[i - 1] == '1' or (s[i - 1] == '2' and s[i] <= '6'):
                cur += prev_prev
            prev, prev_prev = cur, prev

        return prev

        

File: C++/0091-decode-ways.cpp
// Time:  O(n)
// Space: O(1)

class Solution {
public:
    int numDecodings(string s) {
        if (s.empty())
            return 0;
        int prev = 0; // f[n-2]
        int cur = 1; // f[n-1]

        for (int i = 0; i < s.length(); ++ i) {
            if (s[i] == '0')
                cur = 0; // f[n - 1] = 0

            if (i == 0 || !(s[i-1] == '1' || (s[i-1] == '2' && s[i] <= '6')))
                prev = 0;

            int tmp = cur;
            cur = cur + prev; // f[n] = f[n-1] + f[n-2]
            prev = tmp;

        }

        return cur;
    }
};

/*
opt[i] = number of ways to decode string from 0,.......i

opt[i] += opt[i-1], appending "s[i]"" to strings for opt[i-1]

opt[i] += opt[i-2] if s[i-1] == 1 or (s[i-1] == 2 and s[i] <= 6), appending
"s[i-1]s[i]" to strings for opt[i-2]
*/
